hw14/election.py

```diff
@@ -0,0 +1,321 @@
+
+
+def state_edges(election_result_rows):
+    """
+    Given a list of *ElectionDataRow*s, returns *StateEdge*s.
+    The input list has no duplicate *States*;
+    that is, each *State* is represented at most once in the input list.
+    """
+    d = {}
+    for row in election_result_rows:
+        state = row['State']
+        dem = row['Dem']
+        rep = row['Rep']
+
+        d[state] = float(dem) - float(rep)
+    return d
+
+###############################################################################
+# Problem 2: Find the most recent poll row
+###############################################################################
+
+
+def earlier_date(date1, date2):
+    """
+    Given two dates as strings (formatted like "Oct 06 2012"), returns True if
+    date1 is before date2.
+    """
+    return (time.strptime(date1, "%b %d %Y") < time.strptime(date2, "%b %d %Y"))
+
+
+def most_recent_poll_row(poll_rows, pollster, state):
+    """
+    Given a list of *PollDataRow*s, returns the most recent row with the
+    specified *Pollster* and *State*. If no such row exists, returns None.
+    """
+
+    returndata = 'None'
+    checkdate = 'Jan 01 1900'
+    for i in range(len(poll_rows)):
+        if state in poll_rows[i]['State'] and \
+           pollster in poll_rows[i]['Pollster']:
+            if earlier_date(checkdate, poll_rows[i]['Date']):
+                returndata = poll_rows[i]
+        else:
+            continue
+    if returndata == 'None':
+        return None
+    else:
+        return(returndata)
+
+
+###############################################################################
+# Problem 3: Pollster predictions
+###############################################################################
+
+def unique_column_values(rows, column_name):
+    """
+    Given a list of rows and the name of a column (a string),
+    returns a set containing all values in that column.
+    """
+
+    # Make a list of the values corresponding to column names
+    col_list = []
+    for row in rows:
+        col_list.append(row[(column_name)])
+
+    # Reformat/make the list unique using set
+    returndata = set(col_list)
+    return(returndata)
+
+
+def pollster_predictions(poll_rows):
+    """
+    Given a list of *PollDataRow*s, returns *PollsterPredictions*.
+    For a given pollster, uses only the most recent poll for a state.
+    """
+
+    d = {}
+    n = []
+
+    """ I give up. I don't know why this doesn't work."""
+    if len(poll_rows) > 1:
+        for i in range(len(poll_rows) - 1):
+            if poll_rows[i]['Pollster'] == poll_rows[i + 1]['Pollster'] and \
+               poll_rows[i]['State'] == poll_rows[i + 1]['State']:
+                    poll_rows = ([most_recent_poll_row(poll_rows,
+                        poll_rows[i]['Pollster'], poll_rows[i]['State'])])
+
+    edge = state_edges(poll_rows)
+    for pr in poll_rows:
+        edge = state_edges([pr])
+        if pr['Pollster'] in d:
```

Comment:
So let's think about what we're trying to do here. We want to have only the most recent poll for each state for each Pollster. So I think before we even go into this loop we should sort the poll_rows list by date in descending order. And then just add the first instance of each state (which will be the latest). With that check in place, we also need to make sure we're adding in other states for the same pollster. To get you started, here is the sort I wrote, which you are welcome to adapt to your code.

` poll_rows_sorted = sorted(poll_rows, key=lambda row: time.strptime(row['Date'], "%b %d %Y"), reverse=True)`

This sorts poll_rows using the 'Date' value at each row by converting that string into a date object. The `reverse=True` flag at the end says we want this back in descending order, that is from most recent.